### PR TITLE
Comm status filter

### PR DESCRIPTION
**Pull request by fleetwoods**
Adding search by comm status filter to the audit log, and storing the status in dynamo rather than evaluating on the fly.

Dependant on DB migration script being written- do not merge until this is ready!

### CODE REVIEW

File: app/aws/dynamo/Dynamo.scala [hunk 1]
```diff
@@ -66,39 +83,51 @@ object Dynamo {
...
-          Some(
-            queryRequest
-              .withIndexName("customerId-timestamp-index")
-              .addExpressionAttributeNamesEntry("#customerId", "customerId")
-              .addExpressionAttributeNamesEntry("#commName", "commName")
-              .addExpressionAttributeValuesEntry(":customerId", new AttributeValue(cId))
-              .addExpressionAttributeValuesEntry(":commName", new AttributeValue(cName))
-              .withKeyConditionExpression("#customerId = :customerId")
-              .withFilterExpression("#commName = :commName"))
+      val primaryFilter = DbFilter.determineIndex(List(customerId.map(CustomerFilter),
```
Comment (fleetwoods) [inline app/aws/dynamo/Dynamo.scala]: Determine filter rather than index

File: app/aws/dynamo/Dynamo.scala [hunk 2]
```diff
@@ -115,16 +144,62 @@ object Dynamo {
...
+        .addExpressionAttributeValuesEntry(":customerId", new AttributeValue(cStatus))
+        .withFilterExpression("#customerId = :customerId")
+        .addExpressionAttributeNamesEntry("#customerId", "customerId")
+    }
+  }
+
+  // As we can't guarantee the ordering of events we have to check the status in dynamo for certain event updates
+  def statusUpdateExpression(status: CommStatus, traceToken: String, dynamoContext: Context): String = {
+    status match {
+      case CommStatus.Opened => ", #status = :status"
```
Comment (fleetwoods) [inline app/aws/dynamo/Dynamo.scala]: Can't take precedence over link clicked

File: app/models/Comm.scala [hunk 3]
```diff
@@ -22,11 +22,34 @@ import models.EventWrapper.{
...
+
+  val priority: Map[CommStatus, Int] = Map(
+    Opened -> 1,
+    PlatformFailure -> 1,
+    Dropped -> 1,
+    Bounced -> 1,
+    Delivered -> 2,
+    Pending -> 3
+  )
+
```
Comment (fleetwoods) [inline app/models/Comm.scala]: Get rid of Dropped and Bounced, replace with DeliveryFailure - 2
Actioned - 1
Opened - 1

File: app/views/listComms.scala.html [hunk 4]
```diff
@@ -22,14 +22,20 @@
       </div>
     </div>
     <div class="form-group">
+      <label for="commName" class="col-sm-2 control-label">Comm status</label>
+      <div class="col-sm-5">
+        <input id="commName" name="commName" type="text" class="form-control" placeholder="Comm status (e.g Delivered, Opened, Bounced, PlatformFailure)" value="@status">
```
Comment (fleetwoods) [inline app/views/listComms.scala.html]: Dropdown

File: app/views/listComms.scala.html [hunk 5]
```diff
@@ -22,14 +22,20 @@
       </div>
     </div>
     <div class="form-group">
+      <label for="commName" class="col-sm-2 control-label">Comm status</label>
```
Comment (fleetwoods) [inline app/views/listComms.scala.html]: typo- comm status

File: test/ServiceTestIT.scala [hunk 6]
```diff
@@ -79,7 +81,10 @@ class ServiceTestIT
   }
 
   it should "Consume TriggeredV2 events and write appropriate data to dynamodb" taggedAs DockerComposeTag in {
-    val triggered = generate[TriggeredV2].copy(metadata = generate[OrchestratedEmailV2].metadata)
+
+    val templateData = Map("someKey" -> TemplateData(Coproduct[TemplateData.TD]("someValue")))
+    val triggered =
+      generate[TriggeredV2].copy(templateData = templateData, metadata = generate[OrchestratedEmailV2].metadata)
```
Comment (fleetwoods) [inline test/ServiceTestIT.scala]: get rid of the generate[OrchestratedEmailV2] bit

File: test/aws/dynamo/DynamoSpec.scala [hunk 7]
```diff
@@ -180,4 +185,130 @@ class DynamoSpec extends FlatSpec with Matchers with ScalaFutures with TestGener
...
+    }
+  }
+  it should "Should display the correct status for events being written in the wrong order mk2" in {
+    val metadata = generate[Metadata]
+    val triggered = generate[TriggeredV2].copy(metadata = metadata)
+    val delivered = generate[EmailProgressed].copy(metadata = metadata, status = EmailStatus.Delivered)
+    val opened = generate[EmailProgressed].copy(metadata = metadata, status = EmailStatus.Opened)
+    LocalDynamoDB.withTableWithSecondaryIndex(client, tableName)(Seq('traceToken -> S))(secondaryIndices) {
+      Dynamo.writeEvent(opened).run(dynamoContext)
+      val openedResult = Dynamo.search(Some(metadata.customerId), None).run(dynamoContext)
```
Comment (fleetwoods) [inline test/aws/dynamo/DynamoSpec.scala]: Change to Actioned/Interacted

File: test/aws/dynamo/DynamoSpec.scala [hunk 8]
```diff
@@ -102,8 +106,9 @@ class DynamoSpec extends FlatSpec with Matchers with ScalaFutures with TestGener
       val triggeredEvent2 = generate[TriggeredV2]
       Dynamo.writeEvent(triggeredEvent1).run(dynamoContext)
       Dynamo.writeEvent(triggeredEvent2).run(dynamoContext)
-
+      println("wrote event")
```
Comment (fleetwoods) [inline test/aws/dynamo/DynamoSpec.scala]: get rid

File: test/aws/dynamo/DynamoSpec.scala [hunk 9]
```diff
@@ -102,8 +106,9 @@ class DynamoSpec extends FlatSpec with Matchers with ScalaFutures with TestGener
...
       val result = Dynamo.search(Some(triggeredEvent1.metadata.customerId), None).run(dynamoContext)
+      println("Did search")
```
Comment (fleetwoods) [inline test/aws/dynamo/DynamoSpec.scala]: get rid

**Comment by ovo-comms-circleci** [comment]
This PR was successfully deployed to production by [CircleCI build 93](https://circleci.com/gh/ovotech/comms-audit-log/93) at 2017-02-28 11:42 GMT.

**Pull request merged**